### PR TITLE
[L0][Bindless Image] Bug fixes and sampled image support

### PR DESCRIPTION
**Pull request by wenju-he**
This PR fixes some tests on DG2.
* update level-zero loader tag to v1.17.0 which contains definition of
  ZE_IMAGE_BINDLESS_EXP_FLAG_SAMPLED_IMAGE
* Use L0 ImageCopy ext API for copying to/from non-usm
* Use zeCommandListAppendImageCopyRegion for device to device copy
* fix image1DArray copy
* fix pixel size of half type
* support ZE_MEMORY_TYPE_SHARED in bindlessImagesCreate
* Create sampled image from image and sampler descriptors

### CODE REVIEW

**Comment by wenju-he** [comment]
Failure in [Build - Adapters on HW (L0, Debug, clang, clang++)](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#logs) seems unrelated to this pr.
```
Run cmake -B/home/test-user/actions-runner/_work/unified-runtime/unified-runtime/build -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Debug -DUR_ENABLE_TRACING=ON -DUR_DEVELOPER_MODE=ON -DUR_BUILD_TESTS=ON -DUR_BUILD_ADAPTER_L0=ON -DUR_DPCXX=/home/test-user/actions-runner/_work/unified-runtime/unified-runtime/dpcpp_compiler/bin/clang++ -DUR_SYCL_LIBRARY_DIR=/home/test-user/actions-runner/_work/unified-runtime/unified-runtime/dpcpp_compiler/lib   
-- The C compiler identification is Clang 1[4](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:5).0.0
-- The CXX compiler identification is Clang 14.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Python3: /usr/bin/python3.10 (found version "3.10.6") found components: Interpreter 
-- Performing Test USING_LIBSTDCXX
-- Performing Test USING_LIBSTDCXX - Success
-- Fetching sparse source xpti from https://github.com/intel/llvm.git sycl-nightly/20230703
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint: 
hint: 	git config --global init.defaultBranch <name>
hint: 
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint: 
hint: 	git branch -m <name>
Initialized empty Git repository in /home/test-user/actions-runner/_work/unified-runtime/unified-runtime/build/content-xpti/.git/
Switched to a new branch 'main'
error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
error: 1[5](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:6)98 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
error: pathspec 'sycl-nightly/20230703' did not match any file(s) known to git
CMake Error at /usr/share/cmake-3.22/Modules/ExternalProject.cmake:2882 (message):
  No download info given for 'xpti-populate' and its source directory:

   /home/test-user/actions-runner/_work/unified-runtime/unified-runtime/build/content-xpti/xpti
-- Configuring incomplete, errors occurred!
See also "/home/test-user/actions-runner/_work/unified-runtime/unified-runtime/build/CMakeFiles/CMakeOutput.log".

  is not an existing non-empty directory.  Please specify one of:

   * SOURCE_DIR with an existing non-empty directory
   * DOWNLOAD_COMMAND
   * URL
   * GIT_REPOSITORY
   * SVN_REPOSITORY
   * HG_REPOSITORY
   * CVS_REPOSITORY and CVS_MODULE
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/ExternalProject.cmake:371[6](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:7) (_ep_add_download_command)
  CMakeLists.txt:15 (ExternalProject_Add)


-- Configuring incomplete, errors occurred!
See also "/home/test-user/actions-runner/_work/unified-runtime/unified-runtime/build/_deps/xpti-subbuild/CMakeFiles/CMakeOutput.log".

CMake Error at /usr/share/cmake-3.22/Modules/FetchContent.cmake:10[7](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:8)5 (message):
  CMake step for xpti failed: 1
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/FetchContent.cmake:1216:EVAL:2 (__FetchContent_directPopulate)
  /usr/share/cmake-3.22/Modules/FetchContent.cmake:1216 (cmake_language)
  /usr/share/cmake-3.22/Modules/FetchContent.cmake:125[9](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:10) (FetchContent_Populate)
  CMakeLists.txt:[11](https://github.com/oneapi-src/unified-runtime/actions/runs/8415343675/job/23047034417?pr=1470#step:5:12)1 (FetchContent_MakeAvailable)


Error: Process completed with exit code 1.
```

**Review by cppchedy — APPROVED** [review]

**Review by DBDuncan — APPROVED** [review]

**Review by isaacault — APPROVED** [review]

File: source/adapters/level_zero/image.cpp [hunk 1]
```diff
@@ -797,62 +797,70 @@ UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImageCopyExp(
...
       ZE2UR_CALL(zeCommandListAppendMemoryCopyRegion,
-                 (ZeCommandList, pDst, &ZeDstRegion, DstPitch, DstSlicePitch,
-                  pSrc, &ZeSrcRegion, SrcPitch, SrcSlicePitch, ZeEvent,
+                 (ZeCommandList, pDst, &ZeDstRegion, DstRowPitch, DstSlicePitch,
+                  pSrc, &ZeSrcRegion, SrcRowPitch, SrcSlicePitch, ZeEvent,
                   WaitList.Length, WaitList.ZeEventList));
     }
   } else if (imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_DEVICE_TO_HOST) {
+    uint32_t DstRowPitch = hostExtent.width * PixelSizeInBytes;
+    uint32_t DstSlicePitch = DstRowPitch * hostExtent.height;
```
Comment (Seanst98) [inline source/adapters/level_zero/image.cpp:836]: Will there ever be a situation where slice pitch is not `rowPitch * height`? Does it ever make sense to have a different value than that?

**Review by Seanst98 — APPROVED** [review]

File: source/adapters/level_zero/image.cpp (hunk 1 above)
Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:836]: DstRowPitch and DstSlicePitch are properties of host memory. They can differ from rowPitch and slicePitch of the source image.

**Comment by wenju-he** [comment]
@nrspruit could you please review? SYCL PR is at https://github.com/intel/llvm/pull/13926

**Review by nrspruit — APPROVED** [review]
LGTM

**Comment by wenju-he** [comment]
@kbenzie can this PR be added to read-to-merge queue? thank you.

**Comment by kbenzie** [comment]
> @kbenzie can this PR be added to read-to-merge queue? thank you.

I'll add the *ready to merge* label but please do fix the merge conflicts on https://github.com/intel/llvm/pull/13926 so that its ready to go.

**Comment by wenju-he** [comment]
> > @kbenzie can this PR be added to read-to-merge queue? thank you.
> 
> I'll add the _ready to merge_ label but please do fix the merge conflicts on [intel/llvm#13926](https://github.com/intel/llvm/pull/13926) so that its ready to go.

done, merge conflict is resolved. thank you.

**Pull request merged**